File: solutions/go/diamond/3/diamond.go
package diamond

import (
    "errors"
    "strings"
)

func Gen(char byte) (string, error) {
    order := char - 'A'
    if !(0 <= order && order <= 25) {
        return "", errors.New("input is not in valid range")
    }
    side := 2 * order + 1
    buf := make([]byte, side)
    for i := range buf {
        buf[i] = ' '
    }
    diamond := make([]string, side)
    left, right := order, order
    top, bottom := 0, side - 1
    for ch := byte('A'); ch <= char; ch++ {
        buf[left], buf[right] = ch, ch
        diamond[top] = string(buf)
        diamond[bottom] = diamond[top]
        buf[left], buf[right] = ' ', ' '
        left--
        top++
        bottom--
        right++
    }
    return strings.Join(diamond, "\n"), nil
}

File: solutions/csharp/diamond/2/Diamond.cs
public static class Diamond
{
    public static string Make(char target)
    {
        int order = target - 'A', side = 2 * order + 1;
        if (order < 0 || order > 25)
            throw new Exception("target must be between A and Z");
        int left = order, right = order, top = 0, bottom = side - 1;
        char[] buf = new char[side];
        Array.Fill(buf, ' ');
        string[] diamond = new string[side];
        for (char ch = 'A'; ch <= target; ++ch)
        {
            buf[left] = buf[right] = ch;
            diamond[top++] = diamond[bottom--] = new string(buf);
            buf[left--] = buf[right++] = ' ';
        }
        return string.Join('\n', diamond);
    }
}